Adds new API endpoints for search functionality

jacrys opened the pull request:
Adds new API endpoints providing flexible search functionality.

- Introduces `/api/json` for comprehensive search across all entity types with confidence scoring and match details.
- Implements `/api/search` for GeoJSON output, enabling map-based searching with strict/relaxed filtering.
- Implements CORS handling for the new API endpoints.
- Enhances search functionality by incorporating confidence scoring and match examination for better result ranking and insights.
- Includes comprehensive search across buildings, people, census records, media, and rich text content.

jacrys committed:
Merge remote-tracking branch 'upstream/master'

jacrys committed:
Adds new API endpoints for search

Introduces new API endpoints for JSON and GeoJSON search functionality.

This change implements a clean, idiomatic rewrite of the search functionality,
offering unified architecture, simplified hash creation, confidence scoring, and
match examination.

It also adds CORS handling and configuration.

dfurber commented on app/controllers/api/search_controller.rb:
```diff
@@ -0,0 +1,102 @@
+  class SearchController < ApplicationController
+    include EntityBuilders
+    include SearchService
+    include CorsHandler
```
I would consider handling CORS stuff at middleware level with `rack-cors`.

dfurber commented on app/models/building.rb:
```diff
@@ -179,6 +179,36 @@ def self.ransackable_attributes(_auth_object = nil)
     super + %w[street_address]
   end
 
+  def self.as_geojson(buildings)
```
I would make this a service object. I suppose an `as_geojson` instance method could convert a building model to a properties hash. But all the `record.respond_to?` bits make me wonder if it is a building model at all, and if it isn't, what is it doing here in the building model?

dfurber commented on app/controllers/api/search_controller.rb:
```diff
@@ -0,0 +1,102 @@
+
+    def setup_building_compatibility(building, building_hash, year)
+
+      # Set up instance variables exactly like v1
```
Is v1 the `Buildings::MainController`?

dfurber reviewed (commented):
This is interesting but feels purpose built to a frontend to which only you have access. Would you consider adding that too?

Also, I would recommend replacing your "entity builders" of hashes into an app/entities folder that defines PORO (non-AR model) classes, each with a class method that takes a thing, and builds one of them. The building one wants special attention, so you can avoid doing all of that `instance_variable_set` and `define_singleton_method` stuff. (`instance_variable_set` says "I shouldn't be doing this here", and `define_singleton_method` is both bad for performance and says "I am missing a class". That's a start, anyway, I'm curious how this v2 search performs as is.

jacrys commented:
It's not purpose built to a frontend, but rather allows external access to the stored data, decoupling the map display functionality from the model creation and storage, allowing more options for information display than just a map-driven medium.

jacrys commented:
The PORO idea is a duh move and one I am sad to say I didn't think of. I was so focused on getting this out there and working, I missed the forest for the trees. Thanks for pointing that out. I'll go back and reactor that.

jacrys committed:
Refactors API endpoints to use entities

Moves away from ad-hoc hash creation to using entity objects
for API responses.

This change improves code organization, maintainability, and
data consistency across different API endpoints.

Removes CORS handling from controllers and extracts to middleware.
